language/cc/resolve.go

```diff
@@ -188,3 +198,98 @@ func (lang *ccLanguage) resolveImportSpec(c *config.Config, ix *resolve.RuleInde
 
 	return label.NoLabel
 }
+
+func collectStringsAttr(r *rule.Rule, pkgDir, name string) ([]string, error) {
+	// Fast path: plain list of strings in the BUILD file.
+	if ss := r.AttrStrings(name); ss != nil {
+		return ss, nil
+	}
+
+	expr := r.Attr(name) // nil if the attribute is not present
+	if expr == nil {
+		return nil, nil
+	}
+
+	switch e := expr.(type) {
+	case *bzl.ListExpr:
+		return bzl.Strings(e), nil
+
+	case *bzl.CallExpr:
+		id, ok := e.X.(*bzl.Ident)
+		if !ok {
+			break
+		}
+		switch id.Name {
+		case "glob":
+			patterns, excludes := parseGlobCall(e)
+			return expandGlob(pkgDir, patterns, excludes)
+		}
+	}
+	return nil, nil
+}
+
+func parseGlobCall(call *bzl.CallExpr) (patterns, excludes []string) {
+	if len(call.List) > 0 {
+		if lst, ok := call.List[0].(*bzl.ListExpr); ok {
+			patterns = bzl.Strings(lst)
+		}
+	}
+	for idx, arg := range call.List {
+		switch v := arg.(type) {
+		// named argument
+		case *bzl.AssignExpr:
+			lhs, ok := v.LHS.(*bzl.Ident)
+			if !ok {
+				continue
+			}
+			rhs := bzl.Strings(v.RHS)
+			switch lhs.Name {
+			case "include":
+				patterns = rhs
+			case "exclude":
+				excludes = rhs
+			}
+		// positional argument
+		case *bzl.ListExpr:
+			strings := bzl.Strings(arg)
+			switch idx {
+			case 0:
+				patterns = strings
+			case 1:
+				excludes = strings
+			}
+		}
+	}
+	return
+}
+
+func expandGlob(pkgDir string, patterns, excludes []string) ([]string, error) {
+	fsys := os.DirFS(pkgDir)
```

Comment:
Don't use `os.DirFS` here: it allows the `glob` to includes files in Bazel subpackages. We should not include files in directories that have `BUILD` files, including those that Gazelle will generate `BUILD` files for in the future. We also shouldn't include files that Bazel has been told to ignore with `.bazelignore`.

Instead, only consider files in `GenerateArgs.RegularFiles`. That should already be filtered down to the set of regular files in the current package.

Reply:
It was used in `func (*ccLanguage) Imports(c *config.Config, r *rule.Rule, f *rule.File) []resolve.ImportSpec` so no `GenerateArgs` is available. `RegularFiles` might also be not enough because we can define any-subdir glob pattern eg. `dir/**/*.h` so we won't have access to these either.  

Filtering out ignored files in this context, especially  these listed in `.bazelignore` might be non-trivial to implement. Based on these maybe we should not try to implement this feature here if we cannot make it follow the spec.  

One of the goals was indexing rules not managed by gazelle, maybe lazy indexing can replace it with additional explicit configuration

Reply:
In the `GenerateRules` method, you can store the relevant fields of `GenerateArgs` into the `*ccLanguage` receiver. You can confirm which directory you're in using `f.Rel`.

Using the strings from `GenerateRules.RegularFiles`, `.bazelignore`, `# gazelle:exclude` and other directives are already considered, so no need to write new code to handle them.